src/main/java/org/medicmobile/webapp/mobile/ChtExternalAppLauncher.java

```diff
@@ -0,0 +1,378 @@
+package org.medicmobile.webapp.mobile;
+
+import static org.medicmobile.webapp.mobile.JavascriptUtils.safeFormat;
+import static org.medicmobile.webapp.mobile.MedicLog.error;
+import static org.medicmobile.webapp.mobile.MedicLog.trace;
+import static org.medicmobile.webapp.mobile.Utils.getUriFromFilePath;
+
+import android.app.Activity;
+import android.content.Intent;
+import android.graphics.Bitmap;
+import android.graphics.BitmapFactory;
+import android.net.Uri;
+import android.os.Bundle;
+import android.os.ParcelFileDescriptor;
+import android.util.Base64;
+
+import org.json.JSONArray;
+import org.json.JSONException;
+import org.json.JSONObject;
+
+import java.io.ByteArrayOutputStream;
+import java.io.FileInputStream;
+import java.io.InputStream;
+import java.io.Serializable;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.stream.Collector;
+
+public class ChtExternalAppLauncher {
+
+	private ChtExternalAppLauncher() { }
+
+	static Intent createIntent(ChtExternalApp chtExternalApp) {
+		ChtExternalAppIntentBuilder builder = new ChtExternalAppIntentBuilder();
+
+		return builder
+				.setAction(chtExternalApp.getAction())
+				.setCategory(chtExternalApp.getCategory())
+				.setExtras(chtExternalApp.getExtras())
+				.setPackageName(chtExternalApp.getPackageName())
+				.setUri(chtExternalApp.getUri())
+				.setFlags(chtExternalApp.getFlags())
+				.setType(chtExternalApp.getType())
+				.build();
+	}
+
+	static String processResponse(Intent intent, Activity context) {
+		try {
+			if (intent == null || intent.getExtras() == null) {
+				return makeJavaScript(null);
+			}
+
+			JSONObject json = parseBundleToJson(intent.getExtras(), context);
+			return makeJavaScript(json);
+
+		} catch (Exception exception) {
+			error(exception, "ChtExternalAppLauncher :: Problem serialising the intent response");
+			return safeFormat("console.error('Problem serialising the intent response: %s')", exception);
+		}
+	}
+
+	//> PRIVATE
+
+	private static String makeJavaScript(Object response) {
+		String javaScript = "try {" +
+				"const api = window.CHTCore.AndroidApi;" +
+				"if (api && api.resolveCHTExternalAppResponse) {" +
+				"  api.resolveCHTExternalAppResponse(%s);" +
+				"}" +
+				"} catch (error) { " +
+				"  console.error('ChtExternalAppLauncher :: Error on sending intent response to CHT-Core webapp', error);" +
+				"}";
+
+		return safeFormat(javaScript, response);
+	}
+
+	private static JSONObject parseBundleToJson(Bundle bundle, Activity context) {
+		try {
+			JSONObject json = new JSONObject();
+			bundle
+					.keySet()
+					.forEach(key -> setValueInJson(key, json, bundle, context));
+			return json;
+
+		} catch (Exception exception) {
+			error(exception, "ChtExternalAppLauncher :: Problem parsing bundle to json. Bundle=%s", bundle);
+		}
+
+		return null;
+	}
+
+	private static void setValueInJson(String key, JSONObject json, Bundle bundle, Activity context) {
+		try {
+			Object value = bundle.get(key);
```

Comment:
In this type of parsing functions, I tried a bunch of things to reduce the: 
_if string then... if integer then....._ 

At the end decided to use something like: 
`json.put(key, JSONObject.wrap(value));`
or rely on Serializable like: `this.intent.putExtra(key, (Serializable) value);`

Feedback is very welcome!

Reply:
Thanks Josh for looking into this!